backends/simplego/dotgeneral_block.go

```diff
@@ -0,0 +1,323 @@
+	blockDim := 1 << DotGeneralTargetBlockLog2Dim[dtype]
+	if crossSize < blockDim || contractingSize < blockDim {
+		return false
+	}
+
+	// Only pre-block constants or parameters (not computed values)
+	// This ensures we only pay the blocking cost once
+	switch node.opType {
+	case backends.OpTypeParameter, backends.OpTypeConstant:
+		return true
+	default:
+		return false
+	}
+}
+
+// shouldPreBlockRHS determines if the RHS (weights) should be pre-blocked for DotGeneral.
+// This is a specialized version for backward compatibility with 2D weight matrices.
+//
+// Pre-blocking is beneficial when:
+// - RHS is a 2D tensor [K, N] (standard weight layout)
+// - RHS has no batch dimensions (weights are shared across batch)
+// - Single contracting axis on axis 0 for RHS
+// - The dtype is supported for blocking
+// - The matrix is large enough to benefit from blocking (at least one full block in each dimension)
+func shouldPreBlockRHS(rhs *Node, lhsContractingAxes, rhsContractingAxes, rhsBatchAxes []int) bool {
+	// RHS (weights) must be 2D: [K, N]
+	if rhs.shape.Rank() != 2 {
+		return false
+	}
+
+	// RHS must have no batch dimensions (weights are shared across batch)
+	if len(rhsBatchAxes) != 0 {
+		return false
+	}
+
+	// Single contracting axis for RHS
+	if len(rhsContractingAxes) != 1 {
+		return false
+	}
+
+	// RHS contracts on axis 0 (standard [K, N] weight layout)
+	if rhsContractingAxes[0] != 0 {
+		return false
+	}
+
+	// LHS must also have single contracting axis
+	if len(lhsContractingAxes) != 1 {
+		return false
+	}
+
+	// Use the general shouldPreBlock with derived cross/contracting sizes
+	K := rhs.shape.Dimensions[0] // contracting size
+	N := rhs.shape.Dimensions[1] // cross size
+	return shouldPreBlock(rhs, N, K)
+}
+
+// getOrCreateBlockedInputGeneral returns a BlockForDotGeneral node for the given input tensor.
+// If a blocked version already exists (de-duplication), it returns the existing node.
+// Otherwise, it creates a new BlockForDotGeneral node.
+//
+// This is the generalized version that works for both LHS and RHS operands with any shape.
+//
+// Parameters:
+//   - input: the node to block
+//   - contractingAxes, batchAxes: axes from the original tensor shape
+//   - batchSize, crossSize, contractingSize: normalized sizes
+func (b *Builder) getOrCreateBlockedInputGeneral(input *Node,
+	contractingAxes, batchAxes []int,
+	batchSize, crossSize, contractingSize int) *Node {
+
+	dtype := input.shape.DType
+	blockLog2Dim := DotGeneralTargetBlockLog2Dim[dtype]
+	blockedShape := dgCreateBlockedShape(dtype, batchSize, crossSize, contractingSize, blockLog2Dim)
+
+	data := &blockForDotGeneralData{
+		blockLog2Dim:    blockLog2Dim,
+		blockedShape:    blockedShape,
+		batchSize:       batchSize,
+		crossSize:       crossSize,
+		contractingSize: contractingSize,
+		contractingAxes: slices.Clone(contractingAxes),
+		batchAxes:       slices.Clone(batchAxes),
+	}
+
+	// Try to find existing equivalent node via de-duplication
```

Comment:
The pattern: `findDuplicateNode` -> `newNode` -> `registerForDepulication` should be merged into one helper function `b.getOrCreateNewNode()` or something, and used for all ops.

Sry, I just realized that was not what the `dedup` PR did, my bad.